### ISSUE TITLE
Split Interrogate support code "igateruntime" out of interrogatedb

### ISSUE THREAD

**Issue by CFSworks** [panda3d/panda3d#387]
Hey @rdb! I know you wanted to explore doing this; I have something of a PoC about how easy it is to do in the CMake branch here: e501d20927ad1b036e939ba1c204c1481424e324

This would also eliminate pystub as interrogate(_module) no longer depend on the Python library themselves. I also like this option because we can (theoretically) compile igateruntime multiple times, once per supported Python version, and we can support many Python versions with a single Panda build.

Thoughts?

**Comment by rdb** [comment]
This is an improvement over the existing approach.  Though I seem to recall that we decided against this particular approach in #244, since it not only requires us to anticipate and make a version for each ABI but it also requires changes to people's build chains if they use interrogate.  I am sitting on changes that move the type tables to dtoolbase, and if we then make changes to make the rest of py_panda subsist of inline files only, that would obviate the need for a libp3igateruntime.

Are there considerations that make you want to go for a separate support library anyway?  I suppose that the size of the support code has been growing with the addition of `py_wrappers.cxx`, though I don't have the numbers to compare.

If we want to go for a support library anyway, I would suggest that we look into refactoring it to use the [Stable ABI](https://www.python.org/dev/peps/pep-0384/) only, which means we would only need to ship two support libraries: one for Python 2.7, one for Python 3.  I am not sure if this is doable it is (and whether there would be a performance cost) given how extensively the py_panda code relies on playing with the type system internals, but it might be better than having to ship ten copies of the same library.

I'll put my type system changes on a branch shortly, so that they're not just sitting in my patch archive.

**Comment by CFSworks** [comment]
> This is an improvement over the existing approach.  Though I seem to recall that we decided against this particular approach in #244, since it not only requires us to anticipate and make a version for each ABI but it also requires changes to people's build chains if they use interrogate.  I am sitting on changes that move the type tables to dtoolbase, and if we then make changes to make the rest of py_panda subsist of inline files only, that would obviate the need for a libp3igateruntime.

That's fine - I pretty much did this just to unblock myself (needing Python##.dll on the path for Interrogate to work in the build process wasn't really feasible for my own Windows build worker). I still think there should be a p3igateruntime, but it could be a headers-only library of static inline functions, rather than a shared object. (That is to say, the runtime code makes no sense in interrogatedb either way. Only apps interested in .in files should link against p3interrogatedb.)

> Are there considerations that make you want to go for a separate support library anyway?  I suppose that the size of the support code has been growing with the addition of `py_wrappers.cxx`, though I don't have the numbers to compare.

No, just that it made more sense than putting everything in interrogatedb, which made everything _necessarily_ depend on Python. The size, on Windows, of the support library is 72KB (Release), and 817KB (Debug) on Linux. I could see this being a thing people care about for trying to ship really tiny builds of their apps, but I personally don't think anybody would scoff at that small amount of code duplication.

> If we want to go for a support library anyway, I would suggest that we look into refactoring it to use the [Stable ABI](https://www.python.org/dev/peps/pep-0384/) only, which means we would only need to ship two support libraries: one for Python 2.7, one for Python 3.  I am not sure if this is doable it is (and whether there would be a performance cost) given how extensively the py_panda code relies on playing with the type system internals, but it might be better than having to ship ten copies of the same library.

Would we focus on the stable ABI for the Python modules themselves too? At only 72KB per DLL, I think it'd be silly to try to save on many copies of that tiny library but have many copies of the modules (when core.pyd is about a dozen MB, that will add up fast). We'd only be shipping those many copies to developers, and shipping ten copies of the same library to developers would be less of a concern than shipping ten copies in a deployed app to end users.

> I'll put my type system changes on a branch shortly, so that they're not just sitting in my patch archive.

OK - but I am going for 1:1 code parity between `master` and `cmake`, so I won't be removing my hack on the CMake branch until this lands on `master`.

**Comment by rdb** [comment]
I discovered that my code on [`remove-py-code-from-interrogatedb`](https://github.com/panda3d/panda3d/compare/remove-py-code-from-interrogatedb) increases code size in the MSVC build by almost 2 MB for core.pyd.  It probably has to do with an increase in inlining opportunities.  A big point of the shared support code is to reduce code bloat, so that is a shame.

I also performed an experiment with just copy-pasting all the support code into the output of `interrogate_module.cxx`.  This seemed far more effective at keeping code size down, even *reducing* it slightly compared to a pristine master build.

Several ways for us to go ahead with that type of approach are:
1. Create a `py_panda_src.cxx` file, ship it with the Panda3D build in the include directory, and do `#include "py_panda_src.cxx"` inside the generated `xyz_module.cxx` file.
2. Have the build system embed `py_panda.cxx` et al into a large string that is compiled into ` interrogate_module`, which is concatenated into the generated `xyz_module.cxx` file.
3. Instead ship a bunch of static libraries containing the support code, one per Python ABI, and have it linked in when linking the .pyd.

While 1 and 2 don't seem like the best possible coding practices, there are several disadvantages to 3 that make me disprefer it:
* It's a breaking change to projects currently using interrogate
* Users of interrogate would be inherently limited to the Python ABIs that we built with that particular Panda3D distribution; it's suddenly not possible to use interrogate for eg. a new Python version, a debug build of Python, or PyPy, unless we ship support code for literally every possible ABI with our Panda3D builds.

> Would we focus on the stable ABI for the Python modules themselves too? At only 72KB per DLL, I think it'd be silly to try to save on many copies of that tiny library but have many copies of the modules (when core.pyd is about a dozen MB, that will add up fast). We'd only be shipping those many copies to developers, and shipping ten copies of the same library to developers would be less of a concern than shipping ten copies in a deployed app to end users.

The main reason I would advocate the stable ABI if we go with option 3 above is to avoid introducing a limit on the Python ABIs that we can support in any particular distribution of interrogate.  I would much rather let people use the output of interrogate with whatever Python implementation / version / config they like, rather than limit them to the version(s) that their interrogate build happened to be built with.

@cfsworks thoughts?  I'd like to move ahead with landing one approach or the other in 1.10, so that we can move ahead with enabling the same Panda build to support multiple Python versions and obsoleting pystub entirely.  I'm personally leaning towards 1 or 2; I'd need a bit more convincing to go with the support library route, for practicality reasons.

**Comment by CFSworks** [comment]
So to summarize:
1. We **don't** want the support code inlined into individual _igate.cxx files, since it increases the size of the output due to being inlined.
2. We **do** want the support code in the module itself still.
3. We **don't** want to change the invocation of Interrogate or how the output files are handled because it will upset users of Interrogate.

Point two means, no shared runtime library. Point three means, let's not add a static lib.

Overall I think I'd favor a hybrid of 1 and 2: if Interrogate was compiled with the inline-header string(s), use those. If not, fall back on doing #include so as to make development cycles faster. (Implementation-wise this would mean, at compile-time there's a flag that switches between whether the embedded string is `"[contents of py_panda.cxx here]"` or `"#include "/full/path/to/py_panda.cxx"` - Interrogate slips in an embedded string in either case, but the latter doesn't force an Interrogate rebuild)

If I may add a bullet point of my own:
- We **don't** want to add code referencing PyObject to dtoolbase.

It's fine if the type registry lives in dtoolbase, but maybe it should be a language-inspecific type registry. I don't want conflicts if some other implementation of the Python headers (say, PyPy's cpyext) is included before dtoolbase, introducing a conflicting PyObject of its own. To go a step further I might even suggest that the type registry itself shouldn't be part of dtoolbase at all, and it should instead contain a global `void *` that the Interrogate runtime code casts to the type registry singleton. This frees up its dependence on dtool, turning it into more-or-less a pure binding generator.

But maybe I'm just being an architectural purist again and trying to make the boundaries between panda, dtool, and igate clearer than they need to be. :)

**Comment by rdb** [comment]
> Overall I think I'd favor a hybrid of 1 and 2: if Interrogate was compiled with the inline-header string(s), use those. If not, fall back on doing #include so as to make development cycles faster.

Well, if you think it's acceptable to ship a .cxx file in the Panda include directory and `#include` that, then I would prefer to go that route and not have to also maintain something in the build system to turn the .cxx files into a compilable C string.  So I think a hybrid is the worst of both worlds in terms of maintainability.

Unless you are thinking about making interrogate.exe fully self-contained in the long run, by moving `py_panda.h` and the likes to a generated preamble as well?  If we think that is a worthwhile goal, we might as well go the baked-preamble route, with the `#include` as possible development option.

>    * We **don't** want to add code referencing PyObject to dtoolbase.
> 
> 
> It's fine if the type registry lives in dtoolbase, but maybe it should be a language-inspecific type registry.

Right now it's stored as a PyObject pointer, mostly for the benefit of avoiding a cast and automatically having interrogate wrap [this](https://github.com/panda3d/panda3d/blob/e1115bfc1daa0c632f1525e796d7ad6773a9bf9c/dtool/src/dtoolbase/typeHandle.cxx#L156-L167) without need for extension trickery, but this is just an opaque typedef that is in fact compatible with both CPython and PyPy.  Should we make this a void pointer instead?

It's kinda practical that it is stored on the TypeRegistryNode in that we get the benefit of efficiently finding the appropriate type from a type's parents when a subtype doesn't have a corresponding Python wrapper.  That's the main reason I decided to put it in the type registry in the first place; I had previously experimented with putting it in interrogatedb instead.

> To go a step further I might even suggest that the type registry itself shouldn't be part of dtoolbase at all, and it should instead contain a global `void *` that the Interrogate runtime code casts to the type registry singleton. This frees up its dependence on dtool, turning it into more-or-less a pure binding generator.

We already have a strong dependency on dtool in interrogate-generated code, especially when the type system is used.  Though, for interrogating code that doesn't use the type system, it would probably not  be so hard to eliminate this requirement, especially if we go all the way with baked preambles for the support code.

Maybe we can meet in the middle: for the purpose of looking up the Python class corresponding to TypeObject-derived objects in `DtoolInstance_GetPointer`, we continue to store the opaque PyObject pointer (or void pointer if you insist) on the TypedRegistryNode, continuing to get the benefit of having it stored there.
For the named type table, which is for looking up dependent types using their C++ name in a flat global namespace, which we need to do for typed and non-typed classes alike, we instead use a Python dictionary stored in Python's global interpreter state.

That approach would mean there is no dependency on dtool (or even interrogatedb) unless you are using Panda's type system, in which case you are already depending on dtool anyway.

PS.  Once this is done, perhaps we should start building without interrogatedb altogether in ` --optimize=4` builds?  I'm not sure that interrogatedb is very useful in production code.

**Comment by Moguri** [comment]
I don't know how feasible this is, but could interrogatedb be pushed to a plugin instead of being optimized out?

**Comment by CFSworks** [comment]
> Unless you are thinking about making interrogate.exe fully self-contained in the long run, by moving py_panda.h and the likes to a generated preamble as well? If we think that is a worthwhile goal, we might as well go the baked-preamble route, with the #include as possible development option.

That's what I was thinking, yes. I don't like seeing `#include "foo.cxx"` (that would technically be a "header library" and the extension should be .h). But I don't mind having the build process use the bin2c code that ships with pandatool store a preamble inside Interrogate - and when built in development mode, the preamble would just do `#include "/full/path/to/py_panda.{cxx,h}"` so that a recompilation of Interrogate isn't necessary for tweaking the igateruntime/interrogatert files.

That, of course, assumes that those files are edited independently of Interrogate often enough to make that change worthwhile - which may not be the case, so in that circumstance I'd be a strong supporter of option 2.

> Right now it's stored as a PyObject pointer, mostly for the benefit of avoiding a cast and automatically having interrogate wrap this without need for extension trickery, but this is just an opaque typedef that is in fact compatible with both CPython and PyPy. Should we make this a void pointer instead?

As long as we have some way of verifying that the Panda core (not the Interrogate extensions) has no dependency on Python.h then it's actually fine. I was originally worried about Python implementations where PyObject may not be a struct - but it's kind of part of the API that PyObject **must** be a struct, so I think I was being a little too general before.

> PS. Once this is done, perhaps we should start building without interrogatedb altogether in --optimize=4 builds? I'm not sure that interrogatedb is very useful in production code.

> I don't know how feasible this is, but could interrogatedb be pushed to a plugin instead of being optimized out?

As long as there's nothing depending on interrogatedb then we can still build it but not ship/install it (e.g. make it an optional feature in the NSIS installer).

**Comment by rdb** [comment]
Great, thanks for the feedback.  I'll implement the embedding approach first then, but we can always easily switch to using the other approach for development if we feel it has added value (since as you say, it just switches the exact string that is being baked).

> I don't know how feasible this is, but could interrogatedb be pushed to a plugin instead of being optimized out?

The runtime interrogatedb dependency is (after these changes) only used for correlating the databases in the .in files with the produced modules.  This is useful for generating FFI code, generating the API documentation, that sort of thing.  We could probably engineer it so that the dependency is optional, but since I doubt that this is a useful feature in production builds, I tend towards not bothering to build it at all.

> As long as we have some way of verifying that the Panda core (not the Interrogate extensions) has no dependency on Python.h then it's actually fine.

I agree; we shouldn't, and we don't.

**Comment by CFSworks** [comment]
> I agree; we shouldn't, and we don't.

But do we have some way of quickly proving that? Some way that CI can instantly fail a PR by someone who isn't aware that Python doesn't belong in the core? Something that can quickly tell me when there's stray includes like what's going on in physxObject?

Do we just make sure the build system isn't putting any version of Python's include directory on the path? Or do we routinely have CI do a HAVE_PYTHON=OFF build?

**Comment by rdb** [comment]
Makepanda doesn't pass the Python include directory to the compiler unless it's building something explicitly marked as depending on Python, so if it were including Python.h in eg. dtool, the build would simply fail.

(FWIW, I am of mind to just remove the PhysX code entirely rather than fix it up, as the version of the PhysX SDK we are using is no longer supported or maintained, and is very difficult to obtain.)

I do think we should do a HAVE_PYTHON=OFF build in the CI matrix.

**Comment by rdb** [comment]
This is now done in 49b72fb1985a40515eab9a1783d2a9d1bb7d8a47; there is no more dependency from libp3interrogatedb on Python, or from the interrogate tools on pystub.

I've taken advantage of this in 995ba28650e6f1e346552a4a6e8894f69006bf20, which changes makepanda to put all Python-dependent build intermediates in an ABI-tagged subdirectory.  This allows us to invoke makepanda multiple times with different Python versions without needing to rebuild the C++-only stuff, which makes building Panda for multiple Python versions much easier and faster.  If we want, we can now also take the next step of allowing makepanda to build multiple Python versions at the same time, and build installers that support multiple Python versions.

The only thing that still needs pystub is p3dcparse.exe, since the dcparser code uses Python.  I think it's not really a huge deal if this binary happens to either use pystub or depend on Python, but it would be nice to fix this in the long run.  (libRocket and PhysX also still depend on Python, but I think these will not be supported for much longer.)

On a sidenote, I've added a build with `--no-python` to the Travis matrix, since I noticed I had accidentally broken it in a previous commit.

@CFSworks Up to you whether you want to close this issue or leave it open until the code is also physically moved to its own directory.
